larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -326,19 +373,781 @@ StatusCode MipEnergyFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+        "AddMaxFitGapLength", m_addMaxFitGapLength));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "AddRMSLinearFit", m_addRMSLinearFit));
+
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
+        "RatioVariables", m_ratioVariables));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDVertexDistanceFeatureTool::ThreeDVertexDistanceFeatureTool() :
+    m_ratioVariables(false),
+    m_slidingLinearFitWindow(10000)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void ThreeDVertexDistanceFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    //find the neutrino vertex
+    const VertexList *pVertexList = nullptr;
+    (void) PandoraContentApi::GetCurrentList(*pAlgorithm, pVertexList);
+
+    if (pVertexList->empty() || (pVertexList->size() != 1) || (VERTEX_3D != pVertexList->front()->GetVertexType()))
+    {
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+    }
+
+    const Vertex *const nuVertex(pVertexList->front());
+    //find the particle vertex
+    const Vertex *const pVertex = LArPfoHelper::GetVertex(pInputPfo);
+    const CartesianVector nuPosition(nuVertex->GetPosition()), pPosition(pVertex->GetPosition());
+
+    float vertexDistance((nuPosition - pPosition).GetMagnitudeSquared());
+
+    if (m_ratioVariables)
+    {
+        float ratio(-1.f);
+        try
+        {
+            ClusterList clusterList;
+            LArPfoHelper::GetThreeDClusterList(pInputPfo, clusterList);
+            if (1 == clusterList.size())
+            {
+                const Cluster *const pCluster(clusterList.front());
+                const ThreeDSlidingFitResult sliding3DFitResult(pCluster, m_slidingLinearFitWindow, LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+                const TwoDSlidingFitResult slidingFitResult(sliding3DFitResult.GetFirstFitResult());
+                float straightLineLength = (slidingFitResult.GetGlobalMaxLayerPosition() - slidingFitResult.GetGlobalMinLayerPosition()).GetMagnitude();
+                if (straightLineLength > std::numeric_limits<double>::epsilon())
+                    ratio = vertexDistance / straightLineLength;
+            }
```

Comment:
The ratio variables option really does have a large code footprint. If we're using ratio variables, we should think of a way of tidying this; otherwise we can enjoy removing a lot of this stuff!

Reply:
I would/will enjoy removing a lot of stuff